`thrust::optional` fails Clang + older libstdc++ that doesn't have `std::is_trivially_copyable`

When using Clang with older libstdc++ versions that don't have `std::is_trivially_copyable`, `thrust::optional` blows up:

```
swqa@cqa-r28-1:thrust$make thrust.test.async_copy RELEASE=1 USE_CLANGLLVM=/home/swqa/liam/cfe-7.0.0.src/build/bin/clang CXX_STD="c++11"
#### CCBIN         : /home/swqa/liam/cfe-7.0.0.src/build/bin/clang
#### CCBIN VERSION : clang version 7.0.0 (tags/RELEASE_700/final) Target: x86_64-unknown-linux-gnu Thread model: posix InstalledDir: /home/swqa/liam/cfe-7.0.0.src/build/bin
#### CXX_STD       : c++11
Building project testframework release
make ROOTDIR=/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1 POSTCOMP=copy_all_files VULCAN_PROJ_SUBDIR=/internal/build/testframework -C /home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/built/x86_64_Linux_release/thrust/internal/build/testframework -f /home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/internal/build/testframework.mk
ALL_SASS_ARCHITECTURES is [30 35 37 50 52 60 61 70 75]
make[1]: Entering directory `/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/built/x86_64_Linux_release/thrust/internal/build/testframework'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/built/x86_64_Linux_release/thrust/internal/build/testframework'
Building project thrust.test.async_copy release
make ROOTDIR=/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1 POSTCOMP=copy_all_files VULCAN_PROJ_SUBDIR=/../built/x86_64_Linux_agnostic/thrust/mk/thrust.test.async_copy -C /home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/built/x86_64_Linux_release/thrust/../built/x86_64_Linux_agnostic/thrust/mk/thrust.test.async_copy -f /home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/../built/x86_64_Linux_agnostic/thrust/mk/thrust.test.async_copy.mk
ALL_SASS_ARCHITECTURES is [30 35 37 50 52 60 61 70 75]
make[1]: Entering directory `/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/built/x86_64_Linux_release/built/x86_64_Linux_agnostic/thrust/mk/thrust.test.async_copy'
Compiling /home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/testing//async_copy.cu
mkdir -p async_copy.o.keep_dir
/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/bin/x86_64_Linux_release/nvcc        -I/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/testing -I/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/../../../../tools/unix/targets/common/xorg-server/include/ -I/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/cuda/inc -I/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/cuda/tools/cudart -I/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust  -ccbin /home/swqa/liam/cfe-7.0.0.src/build/bin/clang -DSM_30_NVCFG -DSM_35_NVCFG -DSM_37_NVCFG -DSM_50_NVCFG -DSM_52_NVCFG -DSM_60_NVCFG -DSM_61_NVCFG -DSM_70_NVCFG -DSM_75_NVCFG -Xcompiler -msse -Xcompiler -Wall -Xcompiler -Wmultichar -Xcompiler -Wtrigraphs -Xcompiler -Wpointer-arith -Xcompiler -Wformat -Xcompiler -Wreturn-type -Xcompiler -Wno-unused-function -Xcompiler -fno-strict-aliasing -Xcompiler -Wswitch -Xcompiler -Wchar-subscripts -Xcompiler -Wparentheses -Xcompiler -Wcast-align -Xcompiler -Wno-unused-local-typedefs -Xcompiler -O3 -m64 -DCUOS_DISABLE_PRINTS      -Xcompiler "-Wall -Wextra -Werror" -Xcompiler "-Winit-self -Woverloaded-virtual -Wno-cast-align -Wcast-qual -Wno-long-long -Wno-variadic-macros -Wno-unused-function" -Xcompiler "-Wno-unused-parameter" -Xcompiler "-Wno-unneeded-internal-declaration" -Xcompiler "-Wno-noexcept-type" -Wno-deprecated-gpu-targets  --std=c++11   -c /home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/testing//async_copy.cu -o async_copy.o      -gencode arch=compute_30,\"code=sm_30\"     -gencode arch=compute_35,\"code=sm_35\"     -gencode arch=compute_37,\"code=sm_37\"     -gencode arch=compute_50,\"code=sm_50\"     -gencode arch=compute_52,\"code=sm_52\"     -gencode arch=compute_60,\"code=sm_60\"     -gencode arch=compute_61,\"code=sm_61\"     -gencode arch=compute_70,\"code=sm_70\"     -gencode arch=compute_75,\"code=sm_75\"     -gencode arch=compute_75,\"code=compute_75\"
/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(482): error: namespace "std" has no member "is_trivially_copy_constructible"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(482): error: type name is not allowed

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(482): error: the global scope has no "value"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(485): warning: parsing restarts here after previous syntax error

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(489): error: a template argument list is not allowed in a declaration of a primary template

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(518): error: namespace "std" has no member "is_trivially_move_constructible"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(518): error: type name is not allowed

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(518): error: the global scope has no "value"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(521): warning: parsing restarts here after previous syntax error

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(525): error: a template argument list is not allowed in a declaration of a primary template

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(550): error: namespace "std" has no member "is_trivially_copy_assignable"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(550): error: type name is not allowed

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(550): error: the global scope has no "value"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(551): error: namespace "std" has no member "is_trivially_copy_constructible"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(555): warning: parsing restarts here after previous syntax error

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(558): error: a template argument list is not allowed in a declaration of a primary template

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(586): error: namespace "std" has no member "is_trivially_move_constructible"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(586): error: type name is not allowed

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(586): error: the global scope has no "value"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(587): error: namespace "std" has no member "is_trivially_move_assignable"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(590): warning: parsing restarts here after previous syntax error

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(596): error: a template argument list is not allowed in a declaration of a primary template

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(791): error: namespace "thrust::detail" has no member class "optional_move_assign_base"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(791): error: not a class or struct name

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(791): error: class or struct definition is missing

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(884): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(898): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(912): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(927): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(995): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1006): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1017): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1029): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1049): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1061): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1070): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1082): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1091): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1103): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1112): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1124): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1139): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1148): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1157): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1167): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1185): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1196): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1207): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1219): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1229): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1239): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1246): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1253): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1261): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1269): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1276): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1283): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1291): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1300): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1309): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1318): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1328): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1341): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1345): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1353): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1360): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1366): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1376): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1389): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1398): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1408): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1418): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1429): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1439): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1449): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1455): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1470): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1477): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1483): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1501): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1527): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1549): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1563): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1581): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1604): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1612): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1622): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1628): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1633): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1641): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1648): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1653): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1697): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1708): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1719): error: this pragma must immediately precede a declaration

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/optional.h(1727): warning: parsing restarts here after previous syntax error

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/future.inl(318): error: incomplete type is not allowed

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/future.inl(971): error: function "thrust::system::cuda::detail::try_acquire_stream(int, thrust::system::cuda::detail::unique_stream &)" returns incomplete type "thrust::optional<thrust::system::cuda::detail::unique_stream>"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/future.inl(973): error: list-initialization of an object type "thrust::optional<thrust::system::cuda::detail::unique_stream>" is not allowed because the type is incomplete

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/future.inl(978): error: function "thrust::system::cuda::detail::try_acquire_stream(int, thrust::system::cuda::ready_event &)" returns incomplete type "thrust::optional<thrust::system::cuda::detail::unique_stream>"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/future.inl(981): error: list-initialization of an object type "thrust::optional<thrust::system::cuda::detail::unique_stream>" is not allowed because the type is incomplete

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/future.inl(995): error: function "thrust::system::cuda::detail::try_acquire_stream(int, thrust::system::cuda::unique_eager_event &)" returns incomplete type "thrust::optional<thrust::system::cuda::detail::unique_stream>"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/future.inl(1001): error: cannot convert to incomplete class "thrust::optional<thrust::system::cuda::detail::unique_stream>"

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/future.inl(1003): error: list-initialization of an object type "thrust::optional<thrust::system::cuda::detail::unique_stream>" is not allowed because the type is incomplete

/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/future.inl(1040): error: function "thrust::system::cuda::detail::try_acquire_stream(int, thrust::system::cuda::detail::unique_stream &)" returns incomplete type "thrust::optional<thrust::system::cuda::detail::unique_stream>"
          detected during:
            instantiation of "thrust::system::cuda::detail::acquired_stream thrust::system::cuda::detail::acquire_stream_impl(int, std::tuple<Dependencies...> &, thrust::index_sequence<I0, Is...>) [with Dependencies=<thrust::system::cuda::detail::unique_stream>, I0=0UL, Is=<>]"
(1057): here
            instantiation of "thrust::system::cuda::detail::acquired_stream thrust::system::cuda::detail::acquire_stream(int, std::tuple<Dependencies...> &) [with Dependencies=<thrust::system::cuda::detail::unique_stream>]"
(1273): here
            instantiation of "thrust::system::cuda::unique_eager_event thrust::system::cuda::detail::make_dependent_event(std::tuple<Dependencies...> &&) [with Dependencies=<thrust::system::cuda::detail::unique_stream>]"
/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/async/copy.h(114): here
            instantiation of "auto thrust::system::cuda::detail::async_copy_n(FromPolicy &, ToPolicy &, ForwardIt, Size, OutputIt)->std::enable_if<thrust::detail::integral_constant<__nv_bool, <expression>>::value, thrust::system::cuda::unique_eager_event>::type [with FromPolicy=thrust::system::cpp::detail::execution_policy<thrust::system::cpp::detail::tag>, ToPolicy=thrust::cuda_cub::execution_policy<thrust::cuda_cub::tag>, ForwardIt=thrust::detail::normal_iterator<char *>, OutputIt=thrust::detail::normal_iterator<thrust::device_ptr<char>>, Size=std::ptrdiff_t]"
/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/system/cuda/detail/async/copy.h(517): here
            instantiation of "auto thrust::cuda_cub::async_copy(thrust::system::cpp::detail::execution_policy<FromPolicy> &, thrust::cuda_cub::execution_policy<ToPolicy> &, ForwardIt, Sentinel, OutputIt)->decltype((<expression>)) [with FromPolicy=thrust::system::cpp::detail::tag, ToPolicy=thrust::cuda_cub::tag, ForwardIt=thrust::detail::normal_iterator<char *>, Sentinel=thrust::detail::normal_iterator<char *>, OutputIt=thrust::detail::normal_iterator<thrust::device_ptr<char>>]"
/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/thrust/async/copy.h(84): here
            [ 2 instantiation contexts not shown ]
            instantiation of "auto thrust::async::copy_detail::copy_fn::operator()(Args &&...) const->decltype((<expression>)) [with Args=<thrust::detail::normal_iterator<char *>, thrust::detail::normal_iterator<char *>, thrust::detail::normal_iterator<thrust::device_ptr<char>>>]"
/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/testing//async_copy.cu(31): here
            instantiation of "auto invoke_async_copy_fn::operator()(ForwardIt &&, Sentinel &&, OutputIt &&) const->decltype((<expression>)) [with ForwardIt=thrust::detail::normal_iterator<char *>, Sentinel=thrust::detail::normal_iterator<char *>, OutputIt=thrust::detail::normal_iterator<thrust::device_ptr<char>>]"
/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/testing//async_copy.cu(73): here
            instantiation of "void test_async_copy_host_to_device<AsyncCopyCallable>::tester<T>::operator()(std::size_t) [with AsyncCopyCallable=invoke_async_copy_fn, T=char]"
/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/testing/unittest/meta.h(104): here
            instantiation of "void unittest::for_each_type<TypeList, Function, T, i>::operator()(U) [with TypeList=BuiltinNumericTypes, Function=test_async_copy_host_to_device<invoke_async_copy_fn>::tester, T=char, i=0U, U=size_t]"
/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/thrust/testing/unittest/testframework.h(534): here
            instantiation of "void VariableUnitTest<TestName, TypeList>::run() [with TestName=test_async_copy_host_to_device<invoke_async_copy_fn>::tester, TypeList=BuiltinNumericTypes]"

Error limit reached.
100 errors detected in the compilation of "/tmp/tmpxft_0000908c_00000000-14_async_copy.compute_75.cpp1.ii".
Compilation terminated.
make[1]: *** [async_copy.o] Error 1
make[1]: Leaving directory `/home/swqa/p4sw/sw/rel/gpgpu/toolkit/r10.1/built/x86_64_Linux_release/built/x86_64_Linux_agnostic/thrust/mk/thrust.test.async_copy'
make: *** [thrust.test.async_copy] Error 2
swqa@cqa-r28-1:thrust$
```

NVBug 200540293

--- a/thrust/optional.h
+++ b/thrust/optional.h
@@ -59,6 +59,11 @@
   std::has_trivial_copy_constructor<T>::value
 #define THRUST_OPTIONAL_IS_TRIVIALLY_COPY_ASSIGNABLE(T) std::has_trivial_copy_assign<T>::value
 
+// GCC < 5 doesn't provide a way to emulate std::is_trivially_move_*,
+// so don't enable any optimizations that rely on them:
+#define THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_CONSTRUCTIBLE(T) false
+#define THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_ASSIGNABLE(T) false
+
 // This one will be different for GCC 5.7 if it's ever supported
 #define THRUST_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE(T) std::is_trivially_destructible<T>::value
 
@@ -86,13 +91,62 @@ namespace thrust
     thrust::detail::is_trivially_copy_constructible<T>::value
 #define THRUST_OPTIONAL_IS_TRIVIALLY_COPY_ASSIGNABLE(T)                                        \
   std::is_trivially_copy_assignable<T>::value
+#define THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_CONSTRUCTIBLE(T)                                     \
+  std::is_trivially_move_constructible<T>::value
+#define THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_ASSIGNABLE(T)                                        \
+  std::is_trivially_move_assignable<T>::value
 #define THRUST_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE(T) std::is_trivially_destructible<T>::value
 #else
-#define THRUST_OPTIONAL_IS_TRIVIALLY_COPY_CONSTRUCTIBLE(T)                                     \
+
+// To support clang + old libstdc++ without type traits, check for equivalent
+// clang built-ins and use them if present. See note above
+// is_trivially_copyable_impl in
+// thrust/type_traits/is_trivially_relocatable.h for more details.
+
+#ifndef __has_feature
+#define __has_feature(x) 0
+#endif
+
+#if defined(__GLIBCXX__) && __has_feature(is_trivially_constructible)
+#define THRUST_OPTIONAL_IS_TRIVIALLY_COPY_CONSTRUCTIBLE(T) \
+  __is_trivially_constructible(T, T const&)
+#else
+#define THRUST_OPTIONAL_IS_TRIVIALLY_COPY_CONSTRUCTIBLE(T) \
   std::is_trivially_copy_constructible<T>::value
-#define THRUST_OPTIONAL_IS_TRIVIALLY_COPY_ASSIGNABLE(T)                                        \
+#endif
+
+#if defined(__GLIBCXX__) && __has_feature(is_trivially_assignable)
+#define THRUST_OPTIONAL_IS_TRIVIALLY_COPY_ASSIGNABLE(T) \
+  __is_trivially_assignable(T, T const&)
+#else
+#define THRUST_OPTIONAL_IS_TRIVIALLY_COPY_ASSIGNABLE(T) \
   std::is_trivially_copy_assignable<T>::value
-#define THRUST_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE(T) std::is_trivially_destructible<T>::value
+#endif
+
+#if defined(__GLIBCXX__) && __has_feature(is_trivially_constructible)
+#define THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_CONSTRUCTIBLE(T) \
+  __is_trivially_constructible(T, T&&)
+#else
+#define THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_CONSTRUCTIBLE(T) \
+  std::is_trivially_move_constructible<T>::value
+#endif
+
+#if defined(__GLIBCXX__) && __has_feature(is_trivially_assignable)
+#define THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_ASSIGNABLE(T) \
+  __is_trivially_assignable(T, T&&)
+#else
+#define THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_ASSIGNABLE(T) \
+  std::is_trivially_move_assignable<T>::value
+#endif
+
+#if defined(__GLIBCXX__) && __has_feature(is_trivially_destructible)
+#define THRUST_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE(T) \
+  __is_trivially_destructible(T)
+#else
+#define THRUST_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE(T) \
+  std::is_trivially_destructible<T>::value
+#endif
+
 #endif
 
 #if THRUST_CPP_DIALECT > 2011
@@ -511,19 +565,10 @@ struct optional_copy_base<T, false> : optional_operations_base<T> {
   optional_copy_base &operator=(optional_copy_base &&rhs) = default;
 };
 
-// This class manages conditionally having a trivial move constructor
-// Unfortunately there's no way to achieve this in GCC < 5 AFAIK, since it
-// doesn't implement an analogue to std::is_trivially_move_constructible. We
-// have to make do with a non-trivial move constructor even if T is trivially
-// move constructible
-#ifndef THRUST_OPTIONAL_GCC49
-template <class T, bool = std::is_trivially_move_constructible<T>::value>
+template <class T, bool = THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_CONSTRUCTIBLE(T)>
 struct optional_move_base : optional_copy_base<T> {
   using optional_copy_base<T>::optional_copy_base;
 };
-#else
-template <class T, bool = false> struct optional_move_base;
-#endif
 template <class T> struct optional_move_base<T, false> : optional_copy_base<T> {
   using optional_copy_base<T>::optional_copy_base;
 
@@ -578,21 +623,13 @@ struct optional_copy_assign_base<T, false> : optional_move_base<T> {
   operator=(optional_copy_assign_base &&rhs) = default;
 };
 
-// This class manages conditionally having a trivial move assignment operator
-// Unfortunately there's no way to achieve this in GCC < 5 AFAIK, since it
-// doesn't implement an analogue to std::is_trivially_move_assignable. We have
-// to make do with a non-trivial move assignment operator even if T is trivially
-// move assignable
-#ifndef THRUST_OPTIONAL_GCC49
-template <class T, bool = std::is_trivially_destructible<T>::value
-                       &&std::is_trivially_move_constructible<T>::value
-                           &&std::is_trivially_move_assignable<T>::value>
+template <class T,
+          bool = THRUST_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE(T) &&
+                 THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_CONSTRUCTIBLE(T) &&
+                 THRUST_OPTIONAL_IS_TRIVIALLY_MOVE_ASSIGNABLE(T)>
 struct optional_move_assign_base : optional_copy_assign_base<T> {
   using optional_copy_assign_base<T>::optional_copy_assign_base;
 };
-#else
-template <class T, bool = false> struct optional_move_assign_base;
-#endif
 
 template <class T>
 struct optional_move_assign_base<T, false> : optional_copy_assign_base<T> {

